Batch application of style properties

HalfWhitt opened the pull request:
This consists of a few related changes that cut down on the number of calls to `Widget.refresh()` and `Font()`. By my count, when running the core test suite, font creation dropped from 110 to 42, and widget refreshes from 129 to 17.

- `BaseStyle` gains a `batch_apply` context manager in which all calls to `apply()` are aggregated. This is used by:
  - `BaseStyle.update()`
  - Setting or deleting a `directional_property` (and eventually `composite_property`)
  - It doesn't need to be invoked during `__init__`; there's never an applicator present then, so `apply()` is a no-op anyway.
- Apply is split into two methods, an outer `apply` and an inner `_apply`.
  - The outer `apply` is defined in `BaseStyle`, and handles the logic of checking for an applicator and deciding whether to batch the requested name or call `_apply`.
  - Pack (and any other style subclass) now overrides the *inner* one, and does so accepting a set rather than one or more names. The logic then checks for membership of names in this set, rather than iterating through a list. Thus it will never create more than one `Font`, or call `refresh()` more than once per call.
    - This makes the biggest difference when the assignment of a style and an applicator triggers a bare `apply()`, applying all possible style properties.
    - Taking the argument as a set means that the stored `_batched_names` set doesn't need to be expanded into multiple positional arguments and then recreated as a set within the method.
- While setting up the above, I noticed that `validated_property` always calls `apply` on delete, even if the value being deleted is equal to the initial value, so I added a check there (and tested it).

I also combined/parametrized the existing tests for:
- `test_property_with_explicit_const`
- `test_property_with_explicit_value`
- `test_property_with_explicit_none`
- `test_property_with_implicit_default`

I realize this changes the API of Travertino, in terms of what its subclass is supposed to override — but that's the only change. (Except that `apply()` goes back to accepting only a single name, but nothing was written yet that called it with multiple. [Edit: the previous behavior is now maintained, but deprecated.]) And it's still backwards compatible with Toga < 0.5, because existing Toga defines its own `Pack.apply()` method and will completely ignore this; styles that it uses will simply never enter batched mode.

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

HalfWhitt commented:
I also wasn't sure if this would count as a feature or should just be misc.

freakboy3742 commented on travertino/src/travertino/style.py:
```diff
@@ -123,16 +127,56 @@ def copy(self, applicator=None):
+        else:
+            self._apply({name} if name else self._PROPERTIES)
+
+    @cached_property
+    def _batched_names(self):
+        return set()
+
+    @property
+    def _batched_mode(self):
+        return getattr(self, "_batched_mode_stored", False)
+
+    @_batched_mode.setter
+    def _batched_mode(self, value):
+        self._batched_mode_stored = value
+
+        if not value and self._applicator and self._batched_names:
+            self._apply(self._batched_names)
+            self._batched_names.clear()
```
I can see what you're doing here, but I think it's a little over engineered. `_batched_mode` isn't a public API (nor should it be), and it's only used internally in the `batch_apply()` context manager; so abstracting this behind properties makes the code more complex to read in service of a use case that doesn't really exist. 

Similarly, `_batched_names` is a direct attribute of the underlying class; there's no real cost associated with retrieval, so I'm not sure we gain anything from marking it as cached.

HalfWhitt replied:
I see what you mean... however, both of these need to be accessed during `__init__`, in case it sets `margin` (and eventually `font`). We don't have an opportunity to set them on the instance beforehand. We could assign `False` and a blank set to the *class* as a fallback, and then assign "real" ones to the instance in `__post_init__`, but that seems kludgier.

`_batched_names` could return an underlying attribute if present, and if not then create, assign, and return a new `set`... but that would just be spelling out what `cached_property` does.

HalfWhitt replied:
Okay, you're right. 😄

I was thinking about them being needed during `__init__`, but not about the fact that... they're only needed within the `apply` method and the `batched_apply` context manager. I've yanked out the properties and put logic in place to assign them if they're missing.

freakboy3742 commented on travertino/tests/test_style.py:
```diff
@@ -189,176 +189,63 @@ def test_apply(StyleClass):
-    style.apply.assert_called_once_with("implicit")
+    del style[name]
+    assert style[name] == initial
+    style.apply.assert_called_once_with(name)
```
The `explicit_const` version of this test does a second deletion to verify it's a no-op; that wasn't happening for any of the other tests in the parameterized group, but it seems like a test that is worth preserving.

freakboy3742 reviewed (changes requested):
Regarding feature vs misc - this is definitely a feature for me. "Faster" is always a feature :-)

The actual implementation looks pretty good; I've flagged a couple of fairly minor things inline, but they're no more complex than the merge with main that is needed.

There's 2 other issues that aren't flagged inline:

The first is a minor one: given that one of the major optimisations here is the number of updates that font/direction change triggers - it seems like we should have a test that verifies those optimisations are actually in place. Similarly, we should probably have a tests that the "non-layout properties" don't trigger an update, and that layout properties do.

The second is the backwards compatibility issue. I agree with your analysis that a user with an existing subclass of BaseStyle with an `apply()` method won't be impacted by this change (unless the start using `batch_apply`, in which case they'll get the warning about needing `_apply()` - that's fine in my book). However, if someone has upgraded to Toga 0.5 and *is* using `apply('prop1', 'prop2')`, that code will stop working. Looking at the implementation of `apply()` though, I can't see an obvious reason why it couldn't continue to take `*names`, converting that into either a `_batched_names.update(set(names))`, or a `batch_apply()` context if it isn't already being updated? That retains the old `apply(*)` behavior, while also making the apply a batch update.

HalfWhitt commented on core/src/toga/style/pack.py:
```diff
@@ -58,6 +58,21 @@
+    "color",
+    "background_color",
+    "visibility",
+}
```
I agree that would be more readable. — and the same is true for `FONT_PROPERTIES`. The advantage here is that the two sets are created just once, and then only referenced each time the `_apply()` method is called. Do you think that's premature optimization?

freakboy3742 replied:
Yeah - that sounds like premature optimisation to me. It's also something the compiler *could* self-optimize. I don't *think* it will in this case, but given the optimisation work being done in the interpreter at present, it wouldn't surprise me if it *could* be optimised. 

Unless there's *notable* performance benefit, I'm not sure its worth it.

HalfWhitt commented:
> The first is a minor one: given that one of the major optimisations here is the number of updates that font/direction change triggers - it seems like we should have a test that verifies those optimisations are actually in place. Similarly, we should probably have a tests that the "non-layout properties" don't trigger an update, and that layout properties do.

I believe all of those are being tested. In core's `test_style.py`, I modified `test_set_font` to verify that it only sets the font once and triggers one `refresh`, even when multiple font properties are set.

```python
def test_set_font():
    root = ExampleNode(
        "app",
        style=Pack(
            font_family="Roboto",
            font_size=12,
            font_style="normal",
            font_variant="small-caps",
            font_weight="bold",
        ),
    )
    root.style.apply()
    # Should only be called once, despite multiple font-related properties being set.
    root._impl.set_font.assert_called_once_with(
        Font("Roboto", 12, style="normal", variant="small-caps", weight="bold")
    )
    root.refresh.assert_called_once_with()
```

And I added `test_set_multiple_layout_properties` to verify that it only calls `refresh` once when applying multiple layout-relevant properties. It could be more comprehensive. Perhaps it should apply *all* layout-relevant properties?

```python
def test_set_multiple_layout_properties():
    """Setting multiple layout properties at once should only trigger one refresh."""
    root = ExampleNode("app", style=Pack(margin=10, width=100, height=100))
    root.style.apply()
    root.refresh.assert_called_once_with()
```

`test_applicator.py` already has tests that verify layout properties trigger a refresh, and non-layout properties don't:

https://github.com/beeware/toga/blob/9bb3cecf9a13c890ebcf1a523cc09af393d87812/core/tests/style/test_applicator.py#L178-L220

> The second is the backwards compatibility issue. I agree with your analysis that a user with an existing subclass of BaseStyle with an `apply()` method won't be impacted by this change (unless the start using `batch_apply`, in which case they'll get the warning about needing `_apply()` - that's fine in my book). However, if someone has upgraded to Toga 0.5 and _is_ using `apply('prop1', 'prop2')`, that code will stop working. Looking at the implementation of `apply()` though, I can't see an obvious reason why it couldn't continue to take `*names`, converting that into either a `_batched_names.update(set(names))`, or a `batch_apply()` context if it isn't already being updated? That retains the old `apply(*)` behavior, while also making the apply a batch update.

True... My thought process was that the motivation for changing it to `*names` was in order to start moving toward batching style application — but with this implementation that splits `apply` off into `_apply`, that aspect winds up not being relevant. Travertino's API is inward-facing enough that I doubt anyone will have written code using `apply` with multiple names between 0.5.0 and 0.5.1, but it is technically possible. So I guess there are three options:

1. Revert to keeping `*names` as-is, in case we or anyone else finds it useful for something in the future.
2. Revert to keeping it, in case anyone has latched onto it, but add a deprecation notice and remove it later.
3. Remove it now, add a removal notice, and count on it probably not affecting any end users.

HalfWhitt commented:
> Looking at the implementation of `apply()` though, I can't see an obvious reason why it couldn't continue to take `*names`, converting that into either a `_batched_names.update(set(names))`, or a `batch_apply()` context if it isn't already being updated? That retains the old `apply(*)` behavior, while also making the apply a batch update.

If we do keep `*names`, the implementation of `apply('name1', 'name2')` doesn't even have to invoke the `batch_apply` context. As long as it's not currently in batched mode, it would just call `_apply(set(names))`; that's all batching does anyway.

freakboy3742 commented on travertino/src/travertino/style.py:
```diff
@@ -145,30 +144,19 @@ def apply(self, name: str | None = None) -> None:
         if not self._applicator:
             return
 
+        if not hasattr(self, "_batched_names"):
+            self._batched_names = set()
+        if not hasattr(self, "_batched_mode"):
+            self._batched_mode = False
```
Is there any reason to *not* set these in the constructor? That way we *know* they always exist, and don't have to do an existence check every time we do an `apply`.

HalfWhitt replied:
Because unless the user has Python 3.8 or Toga < 0.5, there *is* no constructor. At least not one that we write.

HalfWhitt replied:
This is another one of those cases where everything could work exactly the same at runtime, and be easier to code, if Pack weren't a dataclass — but we wouldn't get IDE autocomplete.

HalfWhitt replied:
It occurs to me that `cached_property` could be useful not only for `batched_names` but for `batched_mode` too. It's only called if the attribute *hasn't* been set, after all...

HalfWhitt replied:
I think there's a better way to handle this: if I rearrange `batch_apply` so that it checks for `_applicator` and aborts before anything else, then the two problematic attributes will never be needed inside `__init__`, so I can set them inside `__post_init__` (and in the fallback `__init__` for when Pack isn't a dataclass). I'll try that when next I'm at a computer.

HalfWhitt replied:
Rearranging pushed 👍

freakboy3742 commented on travertino/src/travertino/style.py:
```diff
@@ -145,30 +144,19 @@ def apply(self, name: str | None = None) -> None:
     @contextmanager
     def batch_apply(self):
-        original = self._batched_mode
+        original = getattr(self, "_batched_mode", False)
```
As above - if we set this in the constructor, we can just use it.

freakboy3742 commented:
> > The first is a minor one: given that one of the major optimisations here is the number of updates that font/direction change triggers - it seems like we should have a test that verifies those optimisations are actually in place. Similarly, we should probably have a tests that the "non-layout properties" don't trigger an update, and that layout properties do.
> 
> I believe all of those are being tested. 

Yup - looks like I missed the significance of some of the tests you added. Thanks for setting me straight.

> And I added `test_set_multiple_layout_properties` to verify that it only calls `refresh` once when applying multiple layout-relevant properties. It could be more comprehensive. Perhaps it should apply _all_ layout-relevant properties?

That was my thinking - if we've got a list of properties that we know should have a specific behavior, we should probably verify that behavior. The actual risk of that part of the code being inadvertently broken isn't *that* high, but the test shouldn't be *that* hard to write either, so we might as well test it.

> So I guess there are three options:
> 
> 1. Revert to keeping `*names` as-is, in case we or anyone else finds it useful for something in the future.
> 2. Revert to keeping it, in case anyone has latched onto it, but add a deprecation notice and remove it later.
> 3. Remove it now, add a removal notice, and count on it probably not affecting any end users.

I think you're right that having an `apply(*names)` API doesn't make a lot of sense when there's a batch update API as well. On that basis, as annoying as it is to have a backwards compatibility API that has only existed for a couple of weeks, I'd say (2) is the way forward.

HalfWhitt commented:
- Font and non-layout property names inlined
- Backwards compatibility shim, test, and removal notice added for `apply()` with multiple arguments
- Layout-batching test now sets every relevant property, all at once

freakboy3742 reviewed (approved):
Awesome - nice cleanups on the `__post_init__` approach.
